src/core/Akka/Actor/GracefulStopSupport.cs

```diff
@@ -15,76 +15,93 @@
+        ///         // Do something after someChild starts being stopped.
+        ///     });
+        /// </code>
         /// </summary>
-        /// <param name="target">TBD</param>
-        /// <param name="timeout">TBD</param>
-        /// <param name="stopMessage">TBD</param>
+        /// <param name="target">The actor to be terminated.</param>
+        /// <param name="timeout">The amount of time we're going to wait for the actor to terminate.</param>
+        /// <param name="stopMessage">A custom message to use to shutdown <see cref="target"/> - by default the other overload uses <see cref="PoisonPill"/>.</param>
         /// <exception cref="TaskCanceledException">
         /// This exception is thrown if the underlying task is <see cref="TaskStatus.Canceled"/>.
         /// </exception>
-        /// <returns>TBD</returns>
-        public static Task<bool> GracefulStop(this IActorRef target, TimeSpan timeout, object stopMessage)
+        /// <returns>A <see cref="Task"/> that will return <c>true</c> if the <see cref="target"/> shuts down within <see cref="timeout"/></returns>
+        public static async Task<bool> GracefulStop(this IActorRef target, TimeSpan timeout, object stopMessage)
         {
-            var internalTarget = target.AsInstanceOf<IInternalActorRef>();
+            if (target is not IInternalActorRef internalTarget)
+                throw new InvalidOperationException(
+                    $"{target} is not an {typeof(IInternalActorRef)} and cannot be death-watched");
 
-            var promiseRef = PromiseActorRef.Apply(internalTarget.Provider, timeout, target, stopMessage.GetType().Name);
+            var promiseRef =
+                PromiseActorRef.Apply(internalTarget.Provider, timeout, target, stopMessage.GetType().Name);
             internalTarget.SendSystemMessage(new Watch(internalTarget, promiseRef));
             target.Tell(stopMessage, ActorRefs.NoSender);
-            return promiseRef.Result.ContinueWith(t =>
+
+            try
             {
-                if (t.Status == TaskStatus.RanToCompletion)
+                var result = await promiseRef.Result;
+                return result switch
                 {
-                    switch (t.Result)
-                    {
-                        case Terminated terminated:
-                            return terminated.ActorRef.Path.Equals(target.Path);
-                        default:
-                            internalTarget.SendSystemMessage(new Unwatch(internalTarget, promiseRef));
-                            return false;
-                        
-                    }
-                }
-                
+                    Terminated terminated => terminated.ActorRef.Path.Equals(target.Path),
+                    _ => false
+                };
+            }
+            catch
+            {
+                // no need to throw here - the returned `false` status does the job just fine
```

Comment:
Behavioral change for `GracefulStop` - no longer throws `TaskCancelledException`s when cancelled. Why? Because returning `false` does this in a much more easily handleable way. The exception adds no value.